src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -257,11 +233,68 @@ private void DrainQueue()
+            //
+            // 2. We specifically only want the updated solution at the start of a request. If the solution provider
+            //    handed out "the right" solution, then a request could ask it for something at the start of processing
+            //    and at the end of processing, and get different results each time.
+
+            // There are multiple possible solutions that we could be interested in, so we need to find the document
+            // first and then get the solution from there. If we're not given a document, this will return the default
+            // solution
+            var (document, solution) = _solutionProvider.GetDocumentAndSolution(queueItem.TextDocument, queueItem.ClientName);
+
+            // Now we can update the solution to represent the LSP view of the world, with any text changes we received
+            solution = await GetSolutionWithReplacedDocumentsAsync(solution, cancellationToken).ConfigureAwait(false);
+
+            // Now we need to get the document again, because if the request is for an open document, the original
+            // one we got might not be up to date
+            if (document is not null)
+            {
+                document = solution.GetDocument(document.Id);
+            }
+
+            DocumentChangeTracker? trackerToUse = null;
+            if (queueItem.MutatesSolutionState)
+            {
+                // Logically, if a mutating request fails we don't want to take its mutations, so giving it the "real" tracker
+                // is a bad idea, but since we tear down the queue for any error anyway, the document tracker will be emptied
+                // and no future requests will be handled, so we don't need to do anything special here.
+                trackerToUse = _documentChangeTracker;
+            }
+
+            return new RequestContext(solution, queueItem.ClientCapabilities, queueItem.ClientName, document, trackerToUse);
+        }
+
+        /// <summary>
+        /// Gets a solution that represents the workspace view of the world (as passed in via the solution parameter)
+        /// but with document text for any open documents updated to match the LSP view of the world. This makes
+        /// the LSP server the source of truth for all document text, but all other changes come from the workspace
+        /// </summary>
+        private async Task<Solution> GetSolutionWithReplacedDocumentsAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            foreach (var trackedDoc in _documentChangeTracker.GetTrackedDocuments())
+            {
+                // We are tracking documents from multiple solutions, so this might not be one we care about
+                if (!solution.ContainsDocument(trackedDoc.Id))
+                {
+                    continue;
+                }
+
+                var text = await trackedDoc.GetTextAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Good question. None of the other LSP handlers do anything special with linked files, so I will look into it.

I will say though that we found the document (and hence document id) from the path (uri) originally, but maybe that doesn't matter ¯\\\_(ツ)_/¯

Reply:
Generally VS passes in a project context with most requests.  The correct document for that context is retrieved here - http://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Extensions/Extensions.cs,123

However - I do not know if they pass us a project context with didOpen/didChange/didClose.  If they don't I expect we need to track all the linked documents?
@jasonmalinowski does the source text for linked documents change between documents (e.g. having defines?)

Maybe we need to track all linked documents here?

Reply:
I don't imagine they're passing a project context on didOpen/didChange/didClose, but even if they were, we probably shouldn't be looking at it, because they won't be sending us events for all linked files. And if a new context came on asynchronously we need to start applying it to all linked copies right away, no matter what.

> @jasonmalinowski does the source text for linked documents change between documents (e.g. having defines?)

The text stays the same, it's the trees that are different. But it's a bit funky in that once you have a Solution snapshot each document can be updated independently. The standard updating code in the workspace will update all copies simultaneously for a text edit, and you do TryApplyChanges we do the merge back.

Reply:
Looked at this and it seems for linked documents, they each have different document IDs, so it should just be a case of finding all documents when a didOpen/Change/Close request comes in, applying the work to each document, and storing individually. Since we already use documentID as the key for tracking, that side of things doesn't need to change.